### PR TITLE
64 build crop and feature extraction pipeline

### PR DESCRIPTION
**Pull request by chris-soelistyo**
- Feature extractor class; has options for per-patch augmentation and transforms (e.g. normalisation). Just extracts bounding boxes, transforms the patches then passes through resnet (or any other extractor) then updates graph.

- Tests for feature extractor; very simply currently, just a forward pass test.

### CODE REVIEW

**Comment by quantumjot** [comment]
@mooniean - can you take the lead on reviewing this one?

File: tests/test_model.py [hunk 1]
```diff
@@ -18,3 +24,47 @@ def test_model_building(input_dims, embedding_dims, output_dims):
...
+        )
+        graph_out = target_out["graph"]
+
+        assert torch.equal(vars["image"], image_out)
+
+        for node in graph_out.nodes.data():
+            x, y = int(node[1][GraphAttrs.NODE_X]), int(
+                node[1][GraphAttrs.NODE_Y]
+            )
+            features = node[1][GraphAttrs.NODE_FEATURES]
```
Comment (mooniean) [inline tests/test_model.py:63]: does the test check also the feature size is the expected? like how the one for resnet152 would be 2048, etc

Comment (chris-soelistyo) [reply tests/test_model.py:63]: Not specifically for resent152, no. Would this be useful? My idea was since it's meant to be model agnostic, there shouldn't be explicit checks for each individual model here, but rather, if we write config tests, or tests for the whole pipeline, then it should check that the extractor output matches the GCN input in this way.

File: grace/models/feature_extractor.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    Normalize,
+    Compose,
+    RandomApply,
+)
+
+from grace.base import GraphAttrs
+from grace.utils.augment_image import RandomEdgeCrop
+
+
+def resnet() -> torch.nn.Module:
```
[the remark this answers is not in the document]
Comment (quantumjot) [reply grace/models/feature_extractor.py]: Personally, I think this was better before - you have no guarantee that you only need to remove the last layer of a model - so I would keep it simple. The module just needs a model that returns some features, that's it

File: grace/models/feature_extractor.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        during training
+    """
+
+    def __init__(
+        self,
+        *,
+        bbox_size: Tuple[int] = (224, 224),
+        model: Callable = classifier_to_feature_extractor(resnet152(ResNet152_Weights)),
+        transforms: Callable = default_transforms,
+        augmentations: Callable = default_augmentations,
```
Comment (quantumjot) [inline grace/models/feature_extractor.py:76]: I suggest you make this a non-optional argument that needs to be added when you set this up

```suggestion
    def __init__(
        self,
        model: Callable,
        *,
        bbox_size: Tuple[int] = (224, 224),
        transforms: Callable = default_transforms,
        augmentations: Callable = default_augmentations,
```

File: grace/models/feature_extractor.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            ]
+            bbox_image = self.transforms(bbox_image)
+
+            if self.training:
+                bbox_image = self.augmentations(bbox_image)
+
+            features = self.model(bbox_image)
+            node[1][GraphAttrs.NODE_FEATURES] = features.squeeze()
+
+        image = image.squeeze()
```
Comment (quantumjot) [inline grace/models/feature_extractor.py:123]: I don't really like this as it has a lot of assumptions. Why not just pass the raw image straight through and make a copy (in the `.ndim < 4` clause) that you use if you need to modify it

File: grace/models/feature_extractor.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        image : torch.Tensor
+            Image of shape (W,H), (C,W,H) or (B,C,W,H)
+        graph : dict
+            Dictionary with key "graph" that accesses the networkx graph
+        """
+
+        if image.ndim < 4:
+            image = image[[None] * (4 - len(image.size()))]
+
+        for node in graph["graph"].nodes.data():
```
Comment (quantumjot) [inline grace/models/feature_extractor.py:102]:
```suggestion
        for node_id, node_attrs in graph["graph"].nodes.data():
```

Comment (quantumjot) [reply grace/models/feature_extractor.py:102]: This get's rid of the [1] indexing later

File: grace/models/feature_extractor.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                - int(self.bbox_size[1] / 2) : y
+                + int(self.bbox_size[1] / 2),
+            ]
+            bbox_image = self.transforms(bbox_image)
+
+            if self.training:
+                bbox_image = self.augmentations(bbox_image)
+
+            features = self.model(bbox_image)
+            node[1][GraphAttrs.NODE_FEATURES] = features.squeeze()
```
Comment (quantumjot) [inline grace/models/feature_extractor.py:121]: Don't use the `node[1]` syntax, better to unpack the tuple in the loop

File: tests/test_model.py [hunk 7]
```diff
@@ -18,3 +24,47 @@ def test_model_building(input_dims, embedding_dims, output_dims):
...
+
+            assert features == model(
+                vars["image"][
+                    x - bbox_size[0] : x + bbox_size[0],
+                    y - bbox_size[1] : y + bbox_size[1],
+                ]
```
Comment (quantumjot) [inline tests/test_model.py:69]: see my earlier comment out using `slice`

**Review by quantumjot — CHANGES_REQUESTED** [review]
A few things to mull over

File: grace/models/feature_extractor.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+            )
+            bbox_image = image[
+                ...,
+                x
+                - int(self.bbox_size[0] / 2) : x
+                + int(self.bbox_size[0] / 2),
+                y
+                - int(self.bbox_size[1] / 2) : y
+                + int(self.bbox_size[1] / 2),
+            ]
```
[the remark this answers is not in the document]
Comment (quantumjot) [reply grace/models/feature_extractor.py:114]: Also - have you checked that x and y are the right way round here?

File: tests/test_model.py (hunk 1 above)
Comment (mooniean) [reply tests/test_model.py:63]: yes, I meant it in an agnostic way, not just specifically for resnet152! something to check that the dimensions are correct and what was expected, but that works out better

Comment (chris-soelistyo) [reply tests/test_model.py:63]: Makes sense! Yeah I'll make sure to add those in the pipeline tests

File: grace/models/feature_extractor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (chris-soelistyo) [reply grace/models/feature_extractor.py:103]: Yeah I think they're actually the other way round; I'll change it now

File: grace/models/feature_extractor.py [hunk 9]
```diff
@@ -85,29 +85,29 @@ def forward(
...
+        image_shape = image.size()
         if image.ndim < 4:
             image = image[[None] * (4 - len(image.size()))]
 
-        for node in graph["graph"].nodes.data():
-            x, y = int(node[1][GraphAttrs.NODE_X]), int(
-                node[1][GraphAttrs.NODE_Y]
+        for node_id, node_attrs in graph["graph"].nodes.data():
+            y, x = int(node_attrs[GraphAttrs.NODE_X]), int(
+                node_attrs[GraphAttrs.NODE_Y]
```
Comment (quantumjot) [inline grace/models/feature_extractor.py:94]: This doesn't seem right somehow NODE_X should be x?

**Review by quantumjot — APPROVED** [review]
Looks good - although still confused by the x, y nomenclature - is it because of the torch dimension order?

File: grace/models/feature_extractor.py (hunk 9 above)
Comment (quantumjot) [reply grace/models/feature_extractor.py:94]: Also, do you need the int here, are it's folded into the slice?
